accounts/tests.py

```diff
@@ -1575,3 +1576,166 @@ def test_bulk_describe_state_closed(self):
         self.client.login(username='testuser', password='testpass')
         resp = self.client.get(reverse('accounts-bulk-describe', args=[bulk.id]))
         self.assertIn('This bulk description process is closed', resp.content)
+
+
+class ReportSpamOffensive(TestCase):
+    """
+    Test the "report spam/offensive" feature available for sound comments, forum posts and private messages.
+    NOTE: for simplicity in  variable names, etc we only refer to the "spam" case.
+    """
+
+    fixtures = ['initial_data', 'sounds']
+
+    def setUp(self):
+        # Create users reporting spam
+        self.reporters = []
+        for i in range(0, settings.USERFLAG_THRESHOLD_FOR_AUTOMATIC_BLOCKING + 1):
+            reporter = User.objects.create_user(username='reporter_{0}'.format(i),
+                                                email='reporter_{0}@example.com'.format(i), password='testpass')
+            self.reporters.append(reporter)
+
+        # Create user posting spam
+        self.spammer = User.objects.create_user(username='spammer', email='spammer@example.com', password='testpass')
+
+    def get_reporter_as_logged_in_user(self, i):
+        user = self.reporters[i]
+        self.client.force_login(user=user)
+        return user
+
+    def __test_report_object(self, flag_type, object):
+
+        # Flag the comment (no email to admins should be sent yet as only one reporter)
+        reporter = self.get_reporter_as_logged_in_user(0)
+        resp = self.client.post(reverse('flag-user', kwargs={'username': self.spammer.username}), data={
+            'object_id': object.id,
+            'flag_type': flag_type,
+        })
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(UserFlag.objects.count(), 1)  # One flag object created
+        self.assertEqual(UserFlag.objects.first().reporting_user, reporter)  # Flag object created by reporter
+        self.assertEqual(len(mail.outbox), 0)  # No email sent
+
+        # Flag the same comment by other users so email is sent
+        for i in range(1, settings.USERFLAG_THRESHOLD_FOR_NOTIFICATION):  # Start at 1 as first flag already done
+            reporter = self.get_reporter_as_logged_in_user(i)
+            resp = self.client.post(reverse('flag-user', kwargs={'username': self.spammer.username}), data={
+                'object_id': object.id,
+                'flag_type': flag_type,
+            })
+            self.assertEqual(resp.status_code, 200)
+            self.assertEqual(UserFlag.objects.count(), i + 1)  # Now we have more flags
+            self.assertEqual(UserFlag.objects.all().order_by('id')[i].reporting_user,
+                             reporter)  # Flag object created by reporter
+
+            if i == settings.USERFLAG_THRESHOLD_FOR_NOTIFICATION - 1:  # Last iteration
```

Comment:
Instead of having all of these asserts and then the conditional, I'd be a little more manual about the checking.
for example, send 1 report, check that there is a UserFlag, check that the outbox is empty, then send all the other reports and check that the email has been triggered.

Reply:
That's in fact what I do, but I agree the last check could be outside the loop and this might make things more clear. I'll update that!